server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -320,20 +320,36 @@ void init() {
                // Remove the listener, but do it on another thread pool to not exhaust the IO thread pool
                addListenerExecutor.submit(() -> {
                   log.debug("Channel disconnected, removing event sender listener for id: " + Util.printArray(listenerId));
-                  cache.removeListener(this);
+                  cache.removeListenerAsync(this);
+                  unblockCommands();
                }));
       }
 
+      private void unblockCommands() {
+         // Have to allow all waiting listeners to proceed
+         for (Iterator<CompletableFuture<Void>> delayedEvent = delayedEvents.iterator(); delayedEvent.hasNext(); ) {
+            CompletableFuture<Void> cf = delayedEvent.next();
+            delayedEvent.remove();
+            cf.complete(null);
```

Comment:
Nitpick, but wouldn't `delayedEvent` be better named as `delayedEventIter` or just `it` etc and then `cf` be `delayedEvent`. It's just `delayedEvent.next()` doesn't sound right to me.

Reply:
Sure, I can add Iter on the end.